modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/leases/LeaseTracker.java

```diff
@@ -255,63 +257,81 @@ private void enqueuePrimaryReplicaEvents(
-        );
-    }
-
     @Override
     public CompletableFuture<ReplicaMeta> awaitPrimaryReplica(
             ReplicationGroupId groupId,
             HybridTimestamp timestamp,
             long timeout,
             TimeUnit unit
     ) {
-        if (!busyLock.enterBusy()) {
-            throw new IgniteInternalException(NODE_STOPPING_ERR, new NodeStoppingException());
-        }
-        try {
+        return inBusyLockAsync(busyLock, () -> {
             ReplicaMeta currentMeta = getCurrentPrimaryReplica(groupId, timestamp);
 
             if (currentMeta != null && clusterNodeResolver.getById(currentMeta.getLeaseholderId()) != null) {
                 return completedFuture(currentMeta);
             }
-        } finally {
-            busyLock.leaveBusy();
-        }
 
-        CompletableFuture<ReplicaMeta> future = new CompletableFuture<>();
+            return awaitPrimaryReplicaImpl(groupId, timestamp, timeout, unit);
+        });
+    }
+
+    private CompletableFuture<ReplicaMeta> awaitPrimaryReplicaImpl(
+            ReplicationGroupId groupId,
+            HybridTimestamp timestamp,
+            long timeout,
+            TimeUnit unit
+    ) {
+        var resultFuture = new CompletableFuture<ReplicaMeta>().orTimeout(timeout, unit);
 
-        awaitPrimaryReplica(groupId, timestamp, future);
+        awaitPrimaryReplicaImpl(groupId, timestamp, resultFuture)
+                .whenComplete(copyStateTo(resultFuture));
 
-        return future
-                .orTimeout(timeout, unit)
+        return resultFuture
                 .exceptionally(e -> {
                     if (e instanceof TimeoutException) {
                         throw new PrimaryReplicaAwaitTimeoutException(groupId, timestamp, leases.leaseByGroupId().get(groupId), e);
+                    } else if (hasCause(e, TrackerClosedException.class)) {
+                        // TrackerClosedException is thrown when trackers are closed on node stop.
+                        throw new CompletionException(new NodeStoppingException(e));
+                    } else {
+                        throw new PrimaryReplicaAwaitException(groupId, timestamp, e);
                     }
-
-                    throw new PrimaryReplicaAwaitException(groupId, timestamp, e);
                 });
     }
 
+    /**
+     * Returns a future that completes when the target primary replica appears.
+     *
+     * <p>{@code timeoutFuture} here is not used for storing the operation result, but rather works in conjunction with the
+     * {@link #awaitPrimaryReplicaImpl(ReplicationGroupId, HybridTimestamp, long, TimeUnit)} method which passes a future that timeouts
+     * after a configurable amount of time. This future is therefore used to stop waiting if the timeout has been reached.
+     */
+    private CompletableFuture<ReplicaMeta> awaitPrimaryReplicaImpl(
+            ReplicationGroupId groupId,
+            HybridTimestamp timestamp,
+            CompletableFuture<?> timeoutFuture
```

Comment:
This approach with returning the future and accepting the result future looks kinda tricky, could you simplify it somehow?